WIP: Edu apply wizard

cehsu opened the pull request:
Connects to department-of-veterans-affairs/vets.gov-team#1493

Description: The PR is intended more for soliciting feedback at this point rather than pulling into master, as it is a proposed change, and has not been formally approved/fully designed.

Instructions for testing this out, please: Navigate to `education/apply-wizard/`

Remaining TODOs
- [x] e2e testing
- [ ] User Testing to determine value of proposed change
- [ ] UI Feedback (May wish to include additional links straight to desired forms)

<img width="829" alt="screen shot 2017-05-17 at 5 32 01 pm" src="https://cloud.githubusercontent.com/assets/16051172/26181688/f152c1aa-3b26-11e7-84a3-bf3d1d489a53.png">
__________________________________________________________________________________________________________
Subsequent questions are revealed once a selection has been made:
<img width="707" alt="screen shot 2017-05-17 at 5 32 08 pm" src="https://cloud.githubusercontent.com/assets/16051172/26181702/1051be94-3b27-11e7-9c3b-ccf3e84f6ce8.png">
__________________________________________________________________________________________________________
Visible questions are toggled once a parent selection has been updated:
<img width="798" alt="screen shot 2017-05-17 at 5 32 15 pm" src="https://cloud.githubusercontent.com/assets/16051172/26181707/1620bd48-3b27-11e7-88ff-715c7d138b57.png">

__________________________________________________________________________________________________________
Application button is revealed once the conclusion of a question path has been reached:
<img width="791" alt="screen shot 2017-05-17 at 5 32 28 pm" src="https://cloud.githubusercontent.com/assets/16051172/26181721/2a49d750-3b27-11e7-9875-dafb6eebf632.png">

Commit:
Relocate and conditionally load script for education/apply/

Commit:
Toggle child question visibility and value based on parent response

Commit:
Map question path using data-* attributes

Commit:
Implement toggle logic

Commit:
Update styling

jbalboni commented on src/js/no-react-entry.js:
```diff
@@ -17,3 +17,7 @@ require('./legacy/mega-menu.js');
 require('./login/login-entry.jsx');
+
+if (location.href.endsWith('education/apply/')) {
+  require('./education-wizard.js');
```
This will keep the code from executing on non-edu pages, but it won't make a separate bundle. However, the file is only 65 lines, I'm not sure it's really worth it to split.

jbalboni commented on src/js/education-wizard.js:
```diff
@@ -0,0 +1,65 @@
+{
```
You shouldn't need these braces around the file.

jbalboni commented on src/js/education-wizard.js:
```diff
@@ -0,0 +1,65 @@
+      return;
+    }
+    const optionsElements = alternateQuestion.querySelectorAll('input');
+    const options = Array.prototype.slice.call(optionsElements);
```
You can use `Array.from()` instead of slice here, since this should now run after we load Babel's ES2105 polyfills.

jbalboni commented on src/js/education-wizard.js:
```diff
@@ -0,0 +1,65 @@
+    const radios = container.querySelectorAll('input');
+    const apply = container.querySelector('#apply-now-button');
+    radios.forEach(radio => {
+      radio.addEventListener('click', () => {
```
This should probably be a change event. With click, this is going to run unnecessarily if someone clicks on an already selected option.

jbalboni commented on src/js/education-wizard.js:
```diff
@@ -0,0 +1,65 @@
+        const nextQuestionElement = container.querySelector(`[data-question=${nextQuestion}]`);
+        const nextQuestionAlternate = nextQuestionElement.dataset.alternate;
+        const nextQuestionAlternateElement = container.querySelector(`[data-question=${nextQuestionAlternate}]`);
+        openState(nextQuestionElement);
```
We probably also want to focus on the newly displayed question label.

jbalboni commented:
@cehsu This looks great so far! One thing we likely need to do is make these changes in a copy of the apply page, so that we could push this to prod and make it easier for @goldenmeanie to user test, but not have it be live immediately.

Second, I noticed that if you fill out a few questions, then go back and change a previous answer, the subsequent questions/submit button don't always show up correctly. You might need to recalculate more/all of the wizard state on each input change.

goldenmeanie commented:
I can run locally if the copy of the page is a pain

jbalboni commented:
I mainly just don't want this PR to sit here for a couple weeks while we test things.

goldenmeanie commented:
Whoa, it's real!

Some feedback:
- Questions need more vertical space between them (replicate in-app style)
- Additional questions should animate in (replicate in-app style)
- Eventual button should be "Apply Now" and width should be content driven, not full width

Commit:
Incorporate feedback

cvalarida commented:
Hey, hey! Looks good. I'm pretty sure it's not actually connected to https://github.com/department-of-veterans-affairs/vets.gov-team/issues/2158, though. That issue is about refactoring the 1995 to use a shared page (which we created after we made the 1995).

Commit:
Remove unused styles

Commit:
Move wizard chagnes to copies

cvalarida commented on src/sass/_shame-wizard.scss:
```diff
@@ -0,0 +1,225 @@
+// This file is a place where you put all the code you're not proud of,
```
Hehe, 'ware the _shame-wizard! He'll replay awkward teenage years again and again and again and...

Maybe rename this to `src/sass/modules/_m-education-wizard.scss`? It's perhaps not the _best_ place for it, but...

cehsu replied:
I like this thought, I will just need to remember to clean up the unused styles in _shame.scss if this change gets approved.

cvalarida commented on src/sass/_shame-wizard.scss:
```diff
@@ -0,0 +1,225 @@
+
+// /education/apply/ uses this expander button for navigation
+// Where else can I put this? Also, it could use a better name.
+.expander-button {
```
Do you need anything above this?

cehsu replied:
The original intention was for this to be a copy to swap in or out, but your idea to create a module is a better way to go about this.

cvalarida reviewed (commented):
I'll hold off on a thorough review until you're ready, but I had a couple things.

Commit:
Create wizard style module

J-Quag commented:
I like where this is headed. :) Keep going! Looking forward to seeing the fleshed out version

jbalboni commented on src/js/education-wizard.js:
```diff
@@ -0,0 +1,61 @@
+  function toggleClass(element, className) {
```
Maybe we should put this file in `src/js/edu-benefits`? We don't have a folder for js for content pages aside from `legacy`, which seems misnamed.

Commit:
Update file structure for easy user testing

Commit:
Create alternate dataset for inputs to simplify hide logic

Commit:
Display warning for non-transfers

Commit:
Lint styles

Commit:
Refactor styles

Commit:
Refactor styles

jbalboni commented on src/js/no-react-entry.js:
```diff
@@ -17,3 +17,7 @@ require('./legacy/mega-menu.js');
 
 require('./login/login-entry.jsx');
+
+// if (location.href.endsWith('education/apply-wizard/')) {
```
I think you can leave this uncommented and do a build type check:

`if (__BUILDTYPE__ !== 'production' && location.href.endsWith('education/apply-wizard/')) {`

Commit:
Create e2e test

Commit:
Resolve style name collisions

Commit:
Hide shame replacement

Commit:
Resolve merge conflict

Commit:
Update test

Commit:
Merge with master

jbalboni reviewed (commented):
Added some more comments. I think this is pretty close to ready. This is what I think is left:

- Clean up shame file
- Add build type check to no-react file
- Add /education/apply-wizard.md to the ignore list in build.js

cvalarida commented on src/js/edu-benefits/education-wizard.js:
```diff
@@ -0,0 +1,73 @@
+    const apply = container.querySelector('#apply-now-button');
+    const warning = container.querySelector('#transfer-warning');
+    radios.forEach(radio => {
+      radio.addEventListener('change', () => {
```
Idle curiosity: @jbalboni Would your suggestion reduce the memory overhead by having only one event listener?

cvalarida commented on test/edu-benefits/edu-apply-wizard.e2e.spec.js:
```diff
@@ -0,0 +1,123 @@
+    // Navigate to application
+    client
+      .click('#apply-now-button');
+    E2eHelpers.expectNavigateAwayFrom(client, '/education/apply-wizard/');
```
My guess is `/education/apply-wizard/` vs `/education/apply-wizard`, but :man_shrugging:

cvalarida commented on src/sass/_shame-wizard.scss:
```diff
@@ -0,0 +1,164 @@
+// hack in there to get it working.
+
+// message while loading react app
+.loading-message {
```
I just did a quick diff between _shame-wizard.scss and _shame.scss. The only difference is _shame.scss also has  the expander stuff used in the current apply page. We can just delete the shame wizard and stick with _shame. (We're currently not even using this file.)

cvalarida commented on src/sass/style.scss:
```diff
@@ -75,7 +75,9 @@
 // ----- VA SHAME ---- //
 
 @import "shame";
+// @import "shame-wizard";
```
Make like a cyberman and DELETE!

cvalarida reviewed (commented):
Mmm...@goldenmeanie, should the apply button show under the warning, leading the user to the 1990e, or should it be hidden?
![image](https://cloud.githubusercontent.com/assets/12970166/26374599/b13ad40c-3fba-11e7-99ff-d8c6a90d952d.png)

Like @jbalboni said, there's a little cleanup left, but it's looking good! I had a few questions, though.

Commit:
Lint scss

cehsu committed:
Update hca-rjsf.scss

Resolve strange linting issue (  3:2  error  Parsing error: Unexpected token) and correspond with master in editor.

Commit:
Remove swap shame from style

Commit:
Add a build type check

Commit:
Add /education/apply-wizard.md to the ignore list in build.js

Commit:
Use event delegation and refine warning toggle conditions

cehsu commented on src/js/edu-benefits/education-wizard.js:
```diff
@@ -0,0 +1,73 @@
+          openState(nextQuestionElement);
+          nextQuestionElement.querySelector('input').focus();
+        }
+        return false;
```
To satisfy: Require Consistent Returns (consistent-return)
https://github.com/department-of-veterans-affairs/vets-website/pull/5577

cvalarida replied:
Oh, eslint, such a whiner. Alright, good enough for me. Alternatively, we could disable the rule with an [inline comment](http://eslint.org/docs/user-guide/configuring.html#disabling-rules-with-inline-comments), but eh. My curiosity is satisfied.

cehsu commented on src/js/edu-benefits/education-wizard.js:
```diff
@@ -0,0 +1,73 @@
+            openState(warning);
+          }
+          apply.href = `/education/apply-for-education-benefits/application/${radio.dataset.selectedForm}/introduction`;
+          return false;
```
To satisfy: Require Consistent Returns (consistent-return)
http://eslint.org/docs/2.0.0/rules/consistent-return

cehsu commented on src/js/edu-benefits/education-wizard.js:
```diff
@@ -0,0 +1,73 @@
+          }
+          if (radio.id === 'create-non-transfer') {
+            openState(warning);
+          }
```
@goldenmeanie explained they should still be able to apply, but that they should also be warned.

cehsu commented on src/js/edu-benefits/education-wizard.js:
```diff
@@ -0,0 +1,73 @@
+          if (!radio.id !== 'create-non-transfer') {
+            closeState(warning);
+          }
+          if (radio.id === 'create-non-transfer') {
```
I have change this so that it also checks if the warning is currently closed.

cehsu commented on src/js/edu-benefits/education-wizard.js:
```diff
@@ -0,0 +1,73 @@
+          if (apply.dataset.state === 'closed') {
+            openState(apply);
+          }
+          if (!radio.id !== 'create-non-transfer') {
```
I have changed this so that it also checks if the warning is currently closed.

Commit:
Lint wizard logic

Commit:
Replace navigate away from logic

goldenmeanie commented:
@cvalarida I think that is a good question. My understanding is that this shouldn't necessarily stop people from applying, it's just that the application won’t go all the way through without the external step. So I think it makes sense to retain the button. Probably should adjust the copy though.

But that is a belief worth validating with stakeholders. I'll ask and get back to y’all

Commit:
Merge branch 'master' into edu-apply-wizard

Commit:
Make test conditional on build type

jbalboni commented on script/build.js:
```diff
@@ -127,6 +127,7 @@ if (options.buildtype === 'production') {
   ignoreList.push('healthcare/rjsf/*');
   ignoreList.push('pensions/application.md');
   ignoreList.push('va-letters/*');
+  ignoreList.push('education/apply-wizard/*');
```
I think this needs to be `education/apply-wizard.md`

jbalboni commented on src/sass/_shame.scss:
```diff
@@ -5,11 +5,12 @@
 
-body .row {
+
+.row {
```
Are all the changes in here necessary? They make me a little nervous, since they'll apply to all pages on vets.gov.

cehsu replied:
All of the changes are of the following type:
-alphabetization of properties
-removing leading zeros from decimal values
-removing units from zero values
-trailing whitespace/missing padding
-double colons for pseudo-elements
-using // instead of /**/ for single line comments
-wrapping strings in double quotes
-removing qualifying elements for class selectors for `body .row` and `a.usa-button-primary` within both `.expander-content-inner` and `.expander-content-closed`, as
--it is unlikely that classes will be assigned outside of the `body` element
--`usa-button-primary` is only ever used for anchor tags within `.expander-content-inner` and `.expander-content-closed`
As a result, these will not affect the styling behavior. These were all recommended by the linter along with some other recommendations (for example, using `!important` and id's were also discouraged, but fixing these seemed likely to necessitate additional changes elsewhere). Please let me know if you prefer to revert any of these.

jbalboni replied:
@cehsu I think we should remove any changes that affect the specificity of styles, since those could result in visual changes. So changes like removing qualifying elements. The other ones are fine.

jbalboni replied:
@cehsu You can make a separate tech debt issue for making the qualifying element changes in that file if you'd like.

jbalboni reviewed (commented):
If we can get rid of the changes in the shame file, I think this is ready to go. If not, we need to do some more testing.

Commit:
Remove wildcard from education/apply-wizard/ in ignoreList

jbalboni reviewed (approved):
This can be merged once the specificity related changes are reverted in the shame scss file.

Commit:
Revert qualifying elements for class selectors

Commit:
Remove linter-required returns and return values

J-Quag commented:
For the link to education about transferring of benefits in the warning message if they select No for Whether the sponsor transferred their benefits:
https://www.dmdc.osd.mil/milconnect/public/faq/Education_Benefits-How_to_Transfer_Benefits

cvalarida commented:
Turned :point_up: into a ticket [here](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/2970).
